File: tools/repomap.py
#!/usr/bin/env python3
"""Generate a compact, signature-only map of the Rust engine source tree.

The goal is to give an AI agent (or a human) a high-signal mental model of the
codebase without paying the token cost of the full source. We emit one block
per file: the relative path followed by its public surface and top-level items
-- module declarations, structs/enums/traits, type aliases, consts/statics,
`impl` headers, and function signatures (name + params + return type, no
bodies). Test modules (`mod tests`) are skipped. Files with nothing to show
are listed with a short tag so the agent still knows the file exists.

Usage (run from the repo root):
    python tools/repomap.py                 # write docs/repo-map.md
    python tools/repomap.py --stdout        # print to stdout instead
    python tools/repomap.py --format json   # write docs/repo-map.jsonl
    python tools/repomap.py --if-stale      # regenerate only if a source file is newer
    python tools/repomap.py --root src --out docs/repo-map.md
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from dataclasses import dataclass, field
from pathlib import Path

# --- Patterns -------------------------------------------------------------
# Visibility prefix (pub / pub(crate) / pub(super) / ...), optional.
_VIS = r"(?:pub(?:\s*\([^)]*\))?\s+)?"

FN = re.compile(rf"^\s*{_VIS}(?:async\s+|unsafe\s+|const\s+|extern\s+\"[^\"]*\"\s+)*fn\s+([A-Za-z_]\w*)")
STRUCT = re.compile(rf"^\s*{_VIS}struct\s+([A-Za-z_]\w*)")
ENUM = re.compile(rf"^\s*{_VIS}enum\s+([A-Za-z_]\w*)")
TRAIT = re.compile(rf"^\s*{_VIS}(?:unsafe\s+)?trait\s+([A-Za-z_]\w*)")
TYPE = re.compile(rf"^\s*{_VIS}type\s+([A-Za-z_]\w*)")
CONST = re.compile(rf"^\s*{_VIS}(?:const|static)\s+(?:mut\s+)?([A-Za-z_]\w*)\s*:\s*([^=;]+?)\s*(?:=|;)")
MOD_DECL = re.compile(rf"^\s*{_VIS}mod\s+([A-Za-z_]\w*)\s*;")
IMPL = re.compile(r"^\s*impl\b(.*)$")
MOD_TESTS = re.compile(r"^\s*(?:#\[cfg\(test\)\]\s*)?mod\s+tests\b")

_LINE_COMMENT = re.compile(r"//.*$")
_STR_DQ = re.compile(r'"(?:\\.|[^"\\])*"')
_STR_SQ = re.compile(r"'(?:\\.|[^'\\])*'")


@dataclass
class FileMap:
    path: str
    signatures: list[str] = field(default_factory=list)


def _norm(text: str) -> str:
    """Collapse all runs of whitespace to single spaces."""
    return re.sub(r"\s+", " ", text).strip()


def _strip_strings(line: str) -> str:
    """Blank out string/char literal contents so brace counting is reliable."""
    return _STR_SQ.sub("''", _STR_DQ.sub('""', line))


def _collect_fn_signature(lines: list[str], i: int) -> str:
    """Join a (possibly multi-line) `fn` definition up to its body/`;`/`where`.

    Returns the normalized signature text starting at the `fn` keyword, e.g.
    `fn run(args: Args) -> Result<()>`. Only paren depth is tracked for the
    stop condition, so generic angle brackets and `->` are left alone.
    """
    buf: list[str] = []
    paren = 0
    bracket = 0
    opened = False
    for j in range(i, len(lines)):
        raw = _LINE_COMMENT.sub("", lines[j])
        seg: list[str] = []
        stop = False
        for ch in raw:
            if ch == "(":
                paren += 1
                opened = True
                seg.append(ch)
            elif ch == ")":
                paren -= 1
                seg.append(ch)
            elif ch == "[":
                bracket += 1
                seg.append(ch)
            elif ch == "]":
                bracket -= 1
                seg.append(ch)
            elif ch == "{" and paren == 0 and bracket == 0 and opened:
                stop = True
                break
            elif ch == ";" and paren == 0 and bracket == 0:
                stop = True
                break
            else:
                seg.append(ch)
        buf.append("".join(seg))
        if stop:
            break
    text = re.split(r"\bwhere\b", " ".join(buf), maxsplit=1)[0]
    return _norm(text)


def extract_signatures(text: str) -> list[str]:
    lines = text.splitlines()
    sigs: list[str] = []
    seen: set[str] = set()

    depth = 0
    skip_until: int | None = None  # brace depth to return to before re-enabling

    for idx, raw in enumerate(lines):
        clean = _strip_strings(_LINE_COMMENT.sub("", raw))

        if skip_until is None and MOD_TESTS.match(raw):
            skip_until = depth  # skip this test module's contents

        if skip_until is None:
            sig: str | None = None

            m = FN.match(raw)
            if m:
                full = _collect_fn_signature(lines, idx)
                after = re.search(rf"\bfn\s+{re.escape(m.group(1))}", full)
                sig = "fn " + m.group(1) + (full[after.end():] if after else "")
                sig = _norm(sig)
            elif (m := STRUCT.match(raw)):
                sig = f"struct {m.group(1)}"
            elif (m := ENUM.match(raw)):
                sig = f"enum {m.group(1)}"
            elif (m := TRAIT.match(raw)):
                sig = f"trait {m.group(1)}"
            elif (m := MOD_DECL.match(raw)):
                sig = f"mod {m.group(1)}"
            elif (m := TYPE.match(raw)):
                sig = f"type {m.group(1)}"
            elif (m := CONST.match(raw)):
                sig = f"const {m.group(1)}: {_norm(m.group(2))}"
            elif (m := IMPL.match(raw)):
                target = _norm(re.split(r"\bwhere\b", m.group(1))[0]).rstrip("{").strip()
                if target:
                    sig = f"impl {target}"

            if sig and sig not in seen:
                seen.add(sig)
                sigs.append(sig)

        depth += clean.count("{") - clean.count("}")
        if skip_until is not None and depth <= skip_until:
            skip_until = None

    return sigs


def build_maps(root: Path, repo_root: Path) -> list[FileMap]:
    maps: list[FileMap] = []
    for path in sorted(root.rglob("*.rs")):
        text = path.read_text(encoding="utf-8", errors="replace")
        rel = path.relative_to(repo_root).as_posix()
        maps.append(FileMap(path=rel, signatures=extract_signatures(text)))
    return maps


def is_stale(root: Path, out_path: Path, generator: Path) -> bool:
    """True if the map is missing or any source file is newer than it.

    Only stats files (no content reads), so the fresh-case check is cheap --
    ideal for running on every prompt submit via a hook.
    """
    if not out_path.exists():
        return True
    out_mtime = out_path.stat().st_mtime
    if generator.exists() and generator.stat().st_mtime > out_mtime:
        return True
    for path in root.rglob("*.rs"):
        if path.stat().st_mtime > out_mtime:
            return True
    return False


def render_text(maps: list[FileMap]) -> str:
    total_sigs = sum(len(m.signatures) for m in maps)
    lines = [
        "# Studio Stud — Repo Map",
        "# Generated by tools/repomap.py · signatures only, no bodies.",
        "# Regenerate after structural changes: python tools/repomap.py",
        f"# {len(maps)} files · {total_sigs} signatures",
        "",
    ]
    for m in maps:
        lines.append(m.path)
        if m.signatures:
            lines.extend(f"  {s}" for s in m.signatures)
        else:
            lines.append("  (no top-level items)")
        lines.append("")
    return "\n".join(lines)


def render_jsonl(maps: list[FileMap]) -> str:
    return "\n".join(
        json.dumps({"path": m.path, "signatures": m.signatures}, separators=(",", ":"))
        for m in maps
    )


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("--root", default="src", help="Source dir to scan (default: src)")
    parser.add_argument("--out", default=None, help="Output file (defaults depend on --format)")
    parser.add_argument("--format", choices=("text", "json"), default="text", help="Output format")
    parser.add_argument("--stdout", action="store_true", help="Print to stdout instead of writing a file")
    parser.add_argument("--if-stale", action="store_true", help="Skip regeneration when no source file is newer than the map")
    parser.add_argument("--quiet", action="store_true", help="Suppress the summary line (for hook/automation use)")
    args = parser.parse_args(argv)

    def log(msg: str) -> None:
        if not args.quiet:
            print(msg)

    repo_root = Path.cwd()
    root = (repo_root / args.root).resolve()
    if not root.exists():
        print(f"error: source dir not found: {root}", file=sys.stderr)
        return 1

    out = args.out or ("docs/repo-map.jsonl" if args.format == "json" else "docs/repo-map.md")
    out_path = repo_root / out

    if args.if_stale and not args.stdout and not is_stale(root, out_path, Path(__file__).resolve()):
        log(f"{out} is up to date")
        return 0

    maps = build_maps(root, repo_root)
    content = render_jsonl(maps) if args.format == "json" else render_text(maps)

    if args.stdout:
        print(content)
        return 0

    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(content + "\n", encoding="utf-8")
    total_sigs = sum(len(m.signatures) for m in maps)
    log(f"Wrote {out} — {len(maps)} files, {total_sigs} signatures")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/repomap.rs
//! Signature-only repo map of the Rust engine.
//!
//! Emits one block per source file: the relative path followed by its public
//! surface and top-level items -- function signatures, structs/enums/traits,
//! type aliases, consts/statics, `impl` headers, and `mod` declarations -- with
//! no bodies. Parsed with `syn` (a real AST), so signatures are accurate
//! regardless of formatting. `mod tests` (and any `#[cfg(test)]` module) is
//! skipped. The goal is a high-signal navigation index an agent can grep
//! instead of reading the whole tree.

use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::SystemTime;

use anyhow::{Context, Result, anyhow};
use serde::Serialize;
use syn::spanned::Spanned;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FileMap {
    pub path: String,
    pub signatures: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RepoMap {
    pub files: Vec<FileMap>,
    pub file_count: usize,
    pub signature_count: usize,
}

/// Collapse whitespace runs and tidy spacing around punctuation so a
/// (possibly multi-line) span reads as a compact one-liner.
fn normalize(text: &str) -> String {
    let collapsed = text.split_whitespace().collect::<Vec<_>>().join(" ");
    collapsed
        .replace("( ", "(")
        .replace(" )", ")")
        .replace(" ,", ",")
        .replace(" ;", ";")
        .replace(",)", ")")
        .trim()
        .to_string()
}

/// Slice the original source for a parsed node's span. Relies on proc-macro2
/// fallback spans (enabled via the `span-locations` feature) which carry byte
/// offsets into the parsed string.
fn span_text(src: &str, span: proc_macro2::Span) -> Option<String> {
    let range = span.byte_range();
    if range.is_empty() {
        return None;
    }
    src.get(range).map(normalize)
}

fn signature_text(src: &str, sig: &syn::Signature) -> String {
    span_text(src, sig.span()).unwrap_or_else(|| format!("fn {}(...)", sig.ident))
}

fn impl_header(src: &str, item: &syn::ItemImpl) -> String {
    let start = item.impl_token.span().byte_range().start;
    let end = item.brace_token.span.open().byte_range().start;
    let header = src
        .get(start..end)
        .map(normalize)
        .unwrap_or_else(|| "impl".to_string());
    let header = header.trim_end_matches('{').trim();
    // Drop a trailing where-clause to keep the header compact.
    let header = header.split(" where ").next().unwrap_or(header).trim();
    header.to_string()
}

fn is_test_module(module: &syn::ItemMod, src: &str) -> bool {
    if module.ident == "tests" {
        return true;
    }
    module.attrs.iter().any(|attr| {
        attr.path().is_ident("cfg")
            && span_text(src, attr.span())
                .map(|t| t.contains("test"))
                .unwrap_or(false)
    })
}

fn push(out: &mut Vec<String>, seen: &mut HashSet<String>, sig: String) {
    if !sig.is_empty() && seen.insert(sig.clone()) {
        out.push(sig);
    }
}

fn collect_items(
    items: &[syn::Item],
    src: &str,
    out: &mut Vec<String>,
    seen: &mut HashSet<String>,
) {
    for item in items {
        match item {
            syn::Item::Fn(f) => push(out, seen, signature_text(src, &f.sig)),
            syn::Item::Struct(s) => push(out, seen, format!("struct {}", s.ident)),
            syn::Item::Enum(e) => push(out, seen, format!("enum {}", e.ident)),
            syn::Item::Type(t) => push(out, seen, format!("type {}", t.ident)),
            syn::Item::Const(c) => {
                let ty = span_text(src, c.ty.span()).unwrap_or_default();
                push(out, seen, format!("const {}: {}", c.ident, ty));
            }
            syn::Item::Static(s) => {
                let ty = span_text(src, s.ty.span()).unwrap_or_default();
                push(out, seen, format!("static {}: {}", s.ident, ty));
            }
            syn::Item::Trait(t) => {
                push(out, seen, format!("trait {}", t.ident));
                for ti in &t.items {
                    if let syn::TraitItem::Fn(m) = ti {
                        push(out, seen, signature_text(src, &m.sig));
                    }
                }
            }
            syn::Item::Impl(im) => {
                push(out, seen, impl_header(src, im));
                for ii in &im.items {
                    if let syn::ImplItem::Fn(m) = ii {
                        push(out, seen, signature_text(src, &m.sig));
                    }
                }
            }
            syn::Item::Mod(m) => {
                if is_test_module(m, src) {
                    continue;
                }
                match &m.content {
                    Some((_, inner)) => collect_items(inner, src, out, seen),
                    None => push(out, seen, format!("mod {}", m.ident)),
                }
            }
            _ => {}
        }
    }
}

/// Extract the ordered, de-duplicated signature list from one Rust source file.
/// Unparseable input yields an empty list (the file is still recorded).
pub fn extract_signatures(src: &str) -> Vec<String> {
    let Ok(file) = syn::parse_file(src) else {
        return Vec::new();
    };
    let mut out = Vec::new();
    let mut seen = HashSet::new();
    collect_items(&file.items, src, &mut out, &mut seen);
    out
}

fn collect_rs_files(dir: &Path, files: &mut Vec<PathBuf>) -> Result<()> {
    for entry in fs::read_dir(dir).with_context(|| format!("read dir {}", dir.display()))? {
        let path = entry?.path();
        if path.is_dir() {
            collect_rs_files(&path, files)?;
        } else if path.extension().is_some_and(|ext| ext == "rs") {
            files.push(path);
        }
    }
    Ok(())
}

pub fn build_repo_map(root: &Path, repo_root: &Path) -> Result<RepoMap> {
    let mut paths = Vec::new();
    collect_rs_files(root, &mut paths)?;
    paths.sort();

    let mut files = Vec::with_capacity(paths.len());
    let mut signature_count = 0;
    for path in paths {
        let text = fs::read_to_string(&path).with_context(|| format!("read {}", path.display()))?;
        let signatures = extract_signatures(&text);
        signature_count += signatures.len();
        let rel = path
            .strip_prefix(repo_root)
            .unwrap_or(&path)
            .to_string_lossy()
            .replace('\\', "/");
        files.push(FileMap {
            path: rel,
            signatures,
        });
    }
    let file_count = files.len();
    Ok(RepoMap {
        files,
        file_count,
        signature_count,
    })
}

pub fn render_text(map: &RepoMap) -> String {
    let mut lines = vec![
        "# Studio Stud — Repo Map".to_string(),
        "# Generated by `studio-stud repo-map` · signatures only, no bodies.".to_string(),
        "# Regenerate after structural changes: studio-stud repo-map".to_string(),
        format!(
            "# {} files · {} signatures",
            map.file_count, map.signature_count
        ),
        String::new(),
    ];
    for file in &map.files {
        lines.push(file.path.clone());
        if file.signatures.is_empty() {
            lines.push("  (no top-level items)".to_string());
        } else {
            lines.extend(file.signatures.iter().map(|s| format!("  {s}")));
        }
        lines.push(String::new());
    }
    lines.join("\n")
}

pub fn render_jsonl(map: &RepoMap) -> String {
    map.files
        .iter()
        .map(|f| serde_json::to_string(f).unwrap_or_default())
        .collect::<Vec<_>>()
        .join("\n")
}

fn newest_source_mtime(root: &Path) -> Option<SystemTime> {
    let mut paths = Vec::new();
    collect_rs_files(root, &mut paths).ok()?;
    paths
        .iter()
        .filter_map(|p| fs::metadata(p).and_then(|m| m.modified()).ok())
        .max()
}

/// True if the map is missing, older than the newest source file, or older than
/// the running binary (a rebuild may change the map's own logic). Stats only --
/// cheap enough to run on every prompt submit.
fn is_stale(root: &Path, out_path: &Path) -> bool {
    let Ok(out_mtime) = fs::metadata(out_path).and_then(|m| m.modified()) else {
        return true;
    };
    if let Some(src_mtime) = newest_source_mtime(root)
        && src_mtime > out_mtime
    {
        return true;
    }
    if let Ok(exe) = std::env::current_exe()
        && let Ok(exe_mtime) = fs::metadata(&exe).and_then(|m| m.modified())
        && exe_mtime > out_mtime
    {
        return true;
    }
    false
}

fn write_atomic(path: &Path, content: &str) -> Result<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let tmp = path.with_extension("tmp");
    fs::write(&tmp, content)?;
    fs::rename(&tmp, path)?;
    Ok(())
}

#[allow(clippy::too_many_arguments)]
pub fn cmd_repo_map(
    root: Option<&Path>,
    out: Option<&Path>,
    as_json: bool,
    to_stdout: bool,
    if_stale: bool,
    quiet: bool,
) -> Result<()> {
    let repo_root = std::env::current_dir()?;
    let root_dir = repo_root.join(root.unwrap_or(Path::new("src")));
    if !root_dir.exists() {
        return Err(anyhow!("source dir not found: {}", root_dir.display()));
    }

    let default_out = if as_json {
        "docs/repo-map.jsonl"
    } else {
        "docs/repo-map.md"
    };
    let out_rel = out
        .map(Path::to_path_buf)
        .unwrap_or_else(|| PathBuf::from(default_out));
    let out_path = repo_root.join(&out_rel);

    if if_stale && !to_stdout && !is_stale(&root_dir, &out_path) {
        if !quiet {
            println!("{} is up to date", out_rel.display());
        }
        return Ok(());
    }

    let map = build_repo_map(&root_dir, &repo_root)?;
    let content = if as_json {
        render_jsonl(&map)
    } else {
        render_text(&map)
    };

    if to_stdout {
        println!("{content}");
        return Ok(());
    }

    write_atomic(&out_path, &format!("{content}\n"))?;
    if !quiet {
        println!(
            "Wrote {} — {} files, {} signatures",
            out_rel.display(),
            map.file_count,
            map.signature_count
        );
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_public_surface() {
        let src = r#"
pub struct Foo;
pub enum Bar { A, B }
pub const N: i64 = 3;
pub type Alias = Vec<u8>;

pub fn free(x: &str) -> Result<()> {
    Ok(())
}

impl Foo {
    pub fn method(
        &self,
        n: usize,
    ) -> [u8; 32] {
        [0; 32]
    }
}

#[cfg(test)]
mod tests {
    fn hidden_helper() {}
}
"#;
        let sigs = extract_signatures(src);
        assert!(sigs.contains(&"struct Foo".to_string()));
        assert!(sigs.contains(&"enum Bar".to_string()));
        assert!(sigs.contains(&"type Alias".to_string()));
        assert!(sigs.iter().any(|s| s.starts_with("const N: i64")));
        assert!(sigs.contains(&"fn free(x: &str) -> Result<()>".to_string()));
        assert!(sigs.contains(&"impl Foo".to_string()));
        // Multi-line method signature is reassembled, array return type intact.
        assert!(sigs.contains(&"fn method(&self, n: usize) -> [u8; 32]".to_string()));
        // Test module contents are skipped.
        assert!(!sigs.iter().any(|s| s.contains("hidden_helper")));
    }

    #[test]
    fn unparseable_file_yields_empty() {
        assert!(extract_signatures("this is not valid rust }{").is_empty());
    }
}
